build.sh

```diff
@@ -35,67 +72,113 @@ do
-  echo '}' >> entities.json
-  perl -Tw .entity-to-dtd.pl < entities-unicode.inc > entities-dtd.url
+  $QUIET || echo "Updating entities database (this always takes a while)...";
+  python .entity-processor.py > $HTML_TEMP/new-entities-unicode.inc;
+  [ -s $HTML_TEMP/new-entities-unicode.inc ] && mv -f $HTML_TEMP/new-entities-unicode.inc $HTML_TEMP/entities-unicode.inc; # otherwise, probably http error, just do it again next time
+  python .entity-processor-json.py > $HTML_TEMP/new-entities-unicode-json.inc;
+  [ -s $HTML_TEMP/new-entities-unicode-json.inc ] && mv -f $HTML_TEMP/new-entities-unicode-json.inc $HTML_TEMP/json-entities-unicode.inc; # otherwise, probably http error, just do it again next time
+  echo '<tbody>' > $HTML_TEMP/entities.inc
+  rm -f $HTML_TEMP/entities-dtd.url $HTML_TEMP/entities*~ $HTML_TEMP/json-*~ $HTML_TEMP/json-entities.inc
+  cat $HTML_TEMP/entities-*.inc | perl -e 'my @lines = <>; print sort { $a =~ m/id="([^"]+?)(-legacy)?"/; $a1 = $1; $a2 = $2; $b =~ m/id="([^"]+?)(-legacy)?"/; $b1 = $1; $b2 = $2; return (lc($a1) cmp lc($b1)) || ($a1 cmp $b1) || ($a2 cmp $b2); } @lines' >> $HTML_TEMP/entities.inc
+  echo '{' > $HTML_OUTPUT/entities.json
+  cat $HTML_TEMP/json-entities-* | sort | perl -e '$/ = undef; $_ = <>; chop, chop, print' >> $HTML_OUTPUT/entities.json
+  echo '' >> $HTML_OUTPUT/entities.json
+  echo '}' >> $HTML_OUTPUT/entities.json
+  perl -Tw .entity-to-dtd.pl < $HTML_TEMP/entities-unicode.inc > $HTML_TEMP/entities-dtd.url
 fi
 
-if [ "$DO_UPDATE" == true ] || [ ! -f caniuse.json ] || [ ! -f w3cbugs.csv ]; then
-  rm -f caniuse.json w3cbugs.csv
+if [ "$DO_UPDATE" == true ] || [ ! -f $HTML_CACHE/caniuse.json ] || [ ! -f $HTML_CACHE/w3cbugs.csv ]; then
+  rm -f $HTML_CACHE/caniuse.json $HTML_CACHE/w3cbugs.csv
   $QUIET || echo "Downloading caniuse data..."
   wget $($VERBOSE || echo "--quiet") \
-    -O caniuse.json --no-check-certificate \
+    -O $HTML_CACHE/caniuse.json --no-check-certificate \
     https://raw.githubusercontent.com/Fyrd/caniuse/master/data.json
-  $QUIET || echo "Downloading list of W3C bugzilla bugs..."
+  $QUIET || echo "Downloading list of W3C bugzilla bugs (can be a wee bit slow)..."
   wget $($VERBOSE || echo "--quiet") \
-    -O w3cbugs.csv \
+    -O $HTML_CACHE/w3cbugs.csv \
     'https://www.w3.org/Bugs/Public/buglist.cgi?columnlist=bug_file_loc,short_desc&query_format=advanced&resolution=---&ctype=csv'
 fi
 
 $QUIET || echo
 $QUIET || echo "Generating spec..."
 $QUIET || echo
-rm -rf source-* .source* .wattsi-*
-perl .pre-process-main.pl < source > .source-expanded-1 || exit
-perl .pre-process-annotate-attributes.pl < .source-expanded-1 > .source-expanded-2 || exit # this one could be merged
-perl .pre-process-tag-omission.pl < .source-expanded-2 > source-whatwg-complete || exit # this one could be merged
+rm -rf $HTML_TEMP/source-* $HTML_TEMP/source* $HTML_TEMP/wattsi-*
+perl .pre-process-main.pl < $HTML_SOURCE/source > $HTML_TEMP/source-expanded-1 || exit 1
```

Comment:
> An alternative to all these || exit 1s is to do set -e

ah yeah, that’s definitely better

> But, it might break some logic. Probably leave it for a separate PR, perhaps one that I make :)

I don’t think it breaks any of the logic, so I’d suggest we go ahead with it.